route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,191 +41,233 @@ namespace route_following_plugin
         
-        ROS_DEBUG_STREAM("Done Loop: approaching_route_end: " << approaching_route_end);
-        if(approaching_route_end){
-            ros::Time end_time = ros::Time::now();
-            if (resp.new_plan.maneuvers.size() > 0) {
-                end_time = resp.new_plan.maneuvers.back().lane_following_maneuver.end_time;
+        double current_downtrack = wm_->routeTrackPos(current_loc_).downtrack;
+        
+        //Return the set of maneuvers which intersect with min_plan_duration
+        int i=0;
+        double planned_time = 0.0;
+        while(planned_time < min_plan_duration_ && i < latest_maneuver_plan_.size()){
+            //Ignore plans for distance already covered
+            if(GET_MANEUVER_PROPERTY(latest_maneuver_plan_[i], end_dist) < current_downtrack){
+                ++i;
+                continue;
             }
+            double maneuver_start_time = GET_MANEUVER_PROPERTY(latest_maneuver_plan_[i], start_time).toSec();
+            double maneuver_end_time = GET_MANEUVER_PROPERTY(latest_maneuver_plan_[i],end_time).toSec();
+            planned_time += maneuver_end_time - maneuver_start_time;
             
-            resp.new_plan.maneuvers.push_back(
-                composeStopandWaitManeuverMessage(current_progress,total_maneuver_length,
-                speed_progress,shortest_path[last_lanelet_index].id(),
-                shortest_path[last_lanelet_index].id(),end_time,time_req_to_stop)); 
+            resp.new_plan.maneuvers.push_back(latest_maneuver_plan_[i]);
+            ++i;
         }
+        
         if(resp.new_plan.maneuvers.size() == 0)
         {
             ROS_WARN_STREAM("Cannot plan maneuver because no route is found");
         }
+        //update plan
+
+        //update start distance of first maneuver and time for all maneuvers
+        setManeuverStartDist(resp.new_plan.maneuvers.front(), current_downtrack);
+        //Update time progress for maneuvers
+        updateTimeProgress(resp.new_plan.maneuvers, ros::Time::now());
+        //update starting speed of first maneuver
+        updateStartingSpeed(resp.new_plan.maneuvers.front(), current_speed_);
+
         return true;
     }
-    void RouteFollowingPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
-    {
-        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+
+    void RouteFollowingPlugin::updateTimeProgress(std::vector<cav_msgs::Maneuver>& maneuvers, ros::Time start_time) const {
+        ros::Time time_progress = start_time;
+        ros::Time prev_time = time_progress;
+
+        for(auto& maneuver : maneuvers){
+            double cur_plus_target = GET_MANEUVER_PROPERTY(maneuver,start_speed) + GET_MANEUVER_PROPERTY(maneuver, end_speed);
+            ros::Duration maneuver_duration;
+            maneuver_duration = ros::Duration((GET_MANEUVER_PROPERTY(maneuver,end_dist) - GET_MANEUVER_PROPERTY(maneuver,start_dist)) / (0.5 * cur_plus_target));
```

Comment:
Calling these GET_MANEUVER_PROPERTY macros twice on the same line is causing a pretty concerning error message on sonar cloud. You should store their output as a local variable 
https://sonarcloud.io/project/issues?id=usdot-fhwa-stol_CARMAPlatform&issues=AXmkZrrBdYWyuZksBLpj&open=AXmkZrrBdYWyuZksBLpj&pullRequest=1272